refactored markdown parser

twerkmeister opened the pull request:
**Proposed changes**:
- refactored markdown parser for greater clarity
- allow -/+/* as list indicators in markdown
- fixes #789 (regex feature support)
- fixes #805 (inline entity synonym support)
- minor doc fix: json training data format needs backslash escaping for regex patterns

**Status (please check what you already did)**:
- [x] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog

twerkmeister commented:
I think travis is caching the training data files that I changed. Not being a member of the org, I cannot restart the build with a fresh cache. Could you do that please, @tmbo?

tmbo commented on data/examples/rasa/demo-rasa.md:
```diff
@@ -49,10 +49,15 @@
 - veggie
+
+## regex:zipcode
+- [0-9]{5}
```
This suggest you can list multiple expressions for the same feature, does that work?

twerkmeister replied:
I am not sure what the intended behavior of the regex featurizer is! If there are multiple patterns defined the md reader generates a json equivalent to:
```json
[{
"name": "zipcode",
"pattern": "pattern1"
},
{
"name": "zipcode",
"pattern": "pattern2"
}]
The regex featurizer doesn't care about the names and just adds the patterns as features it seems (https://github.com/RasaHQ/rasa_nlu/blob/master/rasa_nlu/featurizers/regex_featurizer.py#L81) . So I figured also multiple patterns per name should be fine, but I am not sure what's ultimately intended, since the json format doesn't really incentivize multiple patterns per name

tmbo replied:
👍

tmbo commented on rasa_nlu/utils/md_to_json.py:
```diff
@@ -32,63 +42,71 @@ class MarkdownToJson(object):
+                return name, match.group(1)
+        return None
+
+    def make_json(self):
```
This shouldn't be done in here, as it shouldn't concern the markdown reader (for that matter the class should be renamed as well). The classes only concern should be reading md and creating an instance of `TrainingData`. Something like `TrainingData(trainings_examples, synonyms, regex_features).as_json()` should be used to convert to json.

tmbo commented:
Caching of travis is fixed now

tmbo commented on rasa_nlu/training_data/__init__.py:
```diff
@@ -0,0 +1,21 @@
+        if "value" in s and "synonyms" in s:
+            for synonym in s["synonyms"]:
+                entity_synonyms[synonym] = s["value"]
+    return entity_synonyms
```
newline

tmbo commented on rasa_nlu/converters.py:
```diff
@@ -164,17 +164,11 @@ def load_wit_data(filename):
-    return TrainingData(common_examples, known_synonyms)
+        training_data_unmerged.append(TrainingData.from_markdown(filename))
 
+    return training_data_unmerged[0].merge(training_data_unmerged[1:])
```
fails if empty

tmbo commented on rasa_nlu/training_data/markdown_reader.py:
```diff
@@ -0,0 +1,143 @@
+
+logger = logging.getLogger(__name__)
+
+def create_section_regexes(section_names):
```
start with _ (private)?

tmbo commented on rasa_nlu/training_data/markdown_reader.py:
```diff
@@ -0,0 +1,143 @@
+
+section_regexes = create_section_regexes(available_sections)
+
+def strip_comments(comment_regex, text):
```
start with _ (private)?

tmbo commented on rasa_nlu/training_data/markdown_writer.py:
```diff
@@ -0,0 +1,100 @@
+from rasa_nlu.training_data.markdown_reader import INTENT, SYNONYM, REGEX
+
+
+class MarkdownWriter(object):
```
i think we should put reader and writer into one markdown.py

tmbo commented on rasa_nlu/training_data/training_data.py:
```diff
@@ -78,18 +41,47 @@ def __init__(self,
         self.validate()
 
-    def sanitice_examples(self, examples):
+    def merge(self, others):
```
can we make this `*others` and just handle the list case?

twerkmeister commented:
It's just an intermediate state, that is actually not running rn, should have refactored better...
Am looking for some feedback on the [DataLoader](https://github.com/RasaHQ/rasa_nlu/pull/806/commits/2b299b12677bfeab5b25c8ea4dfbec1553a833d6#diff-dba83ebe2c421de31558dcf5fe0fb864). Is it ok to put so many class/static methods in a class object? I am wondering if I am thinking too much of a scala object. @tmbo

tmbo commented on rasa_nlu/evaluate.py:
```diff
@@ -469,7 +468,7 @@ def run_cv_evaluation(data, n_folds, nlu_config):
 
     if args.mode == "crossvalidation":
-        data = load_data(args.data)
+        data = DataLoader.load(args.data)
```
I don't really like having a class that just consists of static methods (additionally we already have the `TrainingData` class). So from an api perspective I think 
```
from rasa_nlu import training_data

training_data.load()
``` 
is better

twerkmeister replied:
👍 yeah I thought that it was a bit odd

tmbo commented on rasa_nlu/training_data/training_data.py:
```diff
@@ -5,65 +5,29 @@
 from __future__ import print_function
```
I don't think it is a good idea to have a file `training_data` in a module `training_data`

tmbo commented on rasa_nlu/training_data/data_loader.py:
```diff
@@ -0,0 +1,117 @@
+
+    @classmethod
+    def from_wit_file(cls, filename):
+        cls._from_file(WitReader, filename)
```
Do we really need passing around the classes? Why is it better than just calling `WitReader.read(filename)` here?

twerkmeister replied:
Fair enough!

tmbo commented on rasa_nlu/training_data/data_loader.py:
```diff
@@ -0,0 +1,117 @@
+        return reader.read_from_json(js, **kwargs)
+
+    @classmethod
+    def load(cls, resource_name, language='en', fformat=None):
```
This should probably be attached to `TrainingData` class

twerkmeister replied:
I think it could go there, yes. I separated it, because: 1) I felt that data handling finding multiple data files and guessing file formats etc. is a separate concern and 2) the TrainingData class already being quite large itself. Wondering if I should put all the DataLoader functions into the modules __init__.py. Thoughts?

tmbo commented on rasa_nlu/training_data/data_loader.py:
```diff
@@ -0,0 +1,117 @@
+MARKDOWN_FILE_FORMAT = "md"
+
+
+class DataLoader(object):
```
I don't really think we need this class as it really contains no state.

twerkmeister replied:
👍

twerkmeister commented:
Any idea what to do about the failing test: `test_samples` in `test_interpreter.py`? Seems the Mitie extractor finds an entity where it shouldn't find one :D. The training data stats that get logged are in line with the `demo_rasa.json`.
```
training_data.py           212 INFO     Training data stats: 
	- intent examples: 42 (4 distinct intents)
	- Found intents: 'affirm', 'restaurant_search', 'goodbye', 'greet'
	- entity examples: 11 (2 distinct entities)
	- found entities: 'cuisine', 'location'
```

Looks to me like the test is broken, is it possible that updates of the `word_extractor.dat` influence the training and inference?

twerkmeister commented:
Ok, I changed the test to check for specific entities to checking for the entity types instead. Here's why:
1. I changed the training data tests to be much more specific, validating that we get exactly whats in the files. These tests confirm that the component, gets the right data.
2. I manually tested, that the false positive would disappear if we train the MITIE extractor on all examples instead of just those examples with entities.
3. Given the little training data we shouldn't explicitly test for extraction/non-extraction of entities
